insights/parsers/tomcat_virtual_dir_context.py

```diff
@@ -1,46 +1,80 @@
+
+It is especially useful if the Tomcat server is not running.
+
+
+TomcatVirtualDirContextTargeted
+-------------------------------
+
+This is a parser for a command for finding config files in the custom location defined in a command
+line:
+
+``/bin/grep -R -s 'VirtualDirContext' --include '*.xml' {catalina}``
+
+Where ``catalina`` variable is computed as following:
+
+``/bin/ps auxww | awk '/java/ { match($0, "\\-Dcatalina\\.home=([^[:space:]]+)", a); match($0, "\\-Dcatalina\\.base=([^[:space:]]+)", b); if (a[1] != "" || b[1] != "") print a[1] " " b[1] }'``
+
+
+Both parsers detect whether there are any config files which contain VirtualDirContext.
 
 Sample input::
 
     /usr/share/tomcat/conf/server.xml:    <Resources className="org.apache.naming.resources.VirtualDirContext"
 
 Examples::
 
-    >>> shared[TomcatVirtualDirContext].data
+    >>> shared[TomcatVirtualDirContextFallback].data
     {'/usr/share/tomcat/conf/server.xml':
      ['    <Resources className="org.apache.naming.resources.VirtualDirContext"'],
      }
 """
 
 from .. import Parser, parser
-from . import ParseException, SkipException
+from . import SkipException
 
 
-@parser('tomcat_virtual_dir_context')
-class TomcatVirtualDirContext(Parser):
+@parser('tomcat_vdc_fallback')
+class TomcatVirtualDirContextFallback(Parser):
     """
-    Reports whether there is a VirtualDirContext option used in TomCat configuration file.
+    Reports whether there is a VirtualDirContext option used in TomCat configuration file. Looks
+    for the configuration files in default location.
     """
     def __init__(self, *args, **kwargs):
         self.data = {}
-        super(TomcatVirtualDirContext, self).__init__(*args, **kwargs)
+        super(TomcatVirtualDirContextFallback, self).__init__(*args, **kwargs)
 
     def parse_content(self, content):
-        if not content:
-            raise SkipException('VirtualDirContext not used.')
         for line in content:
-            if '/bin/grep: No such file or directory' in line:
-                raise ParseException('grep command not found.')
             try:
                 file_name, file_line = line.split(':', 1)  # Hoping that : is not part of the path
             except ValueError:
-                raise ParseException('Unexpected grep output.')
-            if file_name in self.data:
-                self.data[file_name].append(file_line)
-            else:
-                self.data[file_name] = [file_line]
+                continue  # skip this line
+            if file_name.endswith('.xml'):  # Make sure that nothing else is stored
+                if file_name in self.data:
+                    self.data[file_name].append(file_line)
+                else:
+                    self.data[file_name] = [file_line]
+
+        if self.data == {}:
+            raise SkipException('VirtualDirContext not used.')
+
+
+@parser('tomcat_vdc_targeted')
+class TomcatVirtualDirContextTargeted(TomcatVirtualDirContextFallback):
```

Comment:
Is it certain that TomcatVirtualDirContextFallback being already decorated with `@parser` is not going to cause problems? I remember there used to be such a problem somewhere sometime and the solution was to have one undecorated class and only the leaf children in the inheritance tree could be decorated but IDK if it has been resolved.

Reply:
@skontar I agree with Jakub. There should not be multiple parsers in inheritance. You shoudl add one parent class that inherits from Parser. And two child classes that inherit from the parent class. The child classes should then be decorated with @parser. Look at the iptables parser where I added it like this.

Reply:
Or another example is the 'ps.py' parser.